simln-lib/src/lib.rs

```diff
@@ -1031,257 +1055,191 @@ impl<C: Clock + 'static> Simulation<C> {
+                            capacity
+                        })) => {
+                            let internal_now = SystemTime::now();
+                            let wait_time = match absolute_time.duration_since(internal_now) {
+                                Ok(elapsed) => {
+                                    // the output of duration_since is not perfectly round affecting the waiting time
+                                    // and as consequence the results are not deterministic; for this reason
+                                    // it is necessary to round the output.
+                                    Duration::from_secs(elapsed.as_secs_f64().round() as u64)
+                                },
+                                Err(_e) => {
+                                    Duration::from_secs(0)
                                 }
                             };
+                            let payload = payment_event_payloads.get(&pubkey).ok_or(SimulationError::PaymentGenerationError(
+                                    PaymentGenerationError(format!("executor {} not found", pubkey)),
+                            ))?;
+                            if let Some(c) = payload.executor.payment_generator.payment_count() {
+                                if c == payload.current_count {
+                                    log::info!( "Payment count has been met for {}: {c} payments. Stopping the activity.", payload.executor.source_info);
+                                    continue
+                                }
+                            }
+                            let node = nodes.get(&payload.executor.source_info.pubkey).ok_or(SimulationError::MissingNodeError(format!("Source node not found, {}", payload.executor.source_info.pubkey)))?.clone();
+
+                            // pe: produce events
+                            let pe_shutdown = shutdown.clone();
+                            let pe_clock = clock.clone();
+                            let source = payload.executor.source_info.clone();
+                            let pe_output_sender = output_sender.clone();
+                            // Only proceed with a payment if the amount is non-zero, otherwise skip this round. If we can't get
+                            // a payment amount something has gone wrong (because we should have validated that we can always
+                            // generate amounts), so we exit.
+                            let amount = payload.executor.payment_generator.payment_amount(capacity).map_err(SimulationError::PaymentGenerationError)?;
+                            if amount == 0 {
+                                log::debug!(
+                                    "Skipping zero amount payment for {source} -> {}.", destination
+                                );
+                                generate_payment(&mut heap, pubkey, payload.current_count, &mut payment_event_payloads, SystemTime::now()).await?;
```

Comment:
None of the `?` errors will be caught now, because we're spawning this in a task? I believe the loop will exit but we won't see the error.

I think that we should pull this out into a function and then the spawn callsite is responsible for checking whether it exists with an error and triggering shutdown appropriately if it does (like the tasks spawned in `internal_run`).

Reply:
The `generate_payment` is not part of the `task.spawn`; it is located just before. 
As I understand, the `select!` macro does not spawn tasks :thinking:

Reply:
I mean that we spawn all of this code in a task on line 1088. Any `?` that we use in the block of code that is spawned in the task is just swallowed, so we can silently fail rather than printing out the error.

To try this yourself, add a `return(Err())` anywhere in this select - the program shuts down but we can't see the error.

Reply:
Oh right, I see it now :sweat_smile: